proposed/2021/PackageNamespaces.md

```diff
@@ -0,0 +1,568 @@
+# Package Namespaces
+
+- [Jon Douglas](https://github.com/JonDouglas) & [Nikolche Kolev](https://github.com/nkolev92)
+- Start Date: (2021-02-19)
+- GitHub Issue ([6867](https://github.com/NuGet/Home/issues/6867))
+
+## Summary
+
+<!-- One-paragraph description of the proposal. -->
+NuGet allows access to all package IDs on a package source. However, many .NET developers have needs to filter & consume a subset of the package IDs available on public, private, and local sources. Although NuGet supports a concept of [trusted-signers](https://docs.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-trusted-signers) to trust packages signed by specific authors, users would like to optionally specify package namespaces that their organization and project's software supply chain trusts.
+
+This proposal introduces a concept known as package namespaces, allowing a developer to include or exclude package IDs by specifying package namespaces on their package source(s).
```

Comment:
I'm not familiar enough with the v3 http protocol, but does the package feed expose the registered author of the package? If so, we'd much prefer the option to whitelist by author instead of by namespace. 

Rationale: With the PackageReference mechanism and .g.props auto-importing certain named files in the package, package authors are granted permission to run arbitrary code during the user's build. Given that, a given author could easily impersonate other packages by rewriting those packages on disk.

When whitelisting a namespace, I'm effectively granting permission to all authors that have the ability to publish into that namespace to impersonate any package. So why not cut out the intermediary and list authors specifically, if feasible?

Secondly, users are likely to use wildcards without first checking that it properly is scoped to an author. Do I know for certain that nuget.org will only let NewtonSoft publish to the NewtonSoft package namespace? I know many users that would jump to the assumption that nuget.org would implement that restriction.

Reply:
There's an interesting relationship here:

ID prefixes have a 1 to many relationship with owners.

An allow list of owners/maintainers is interesting. You may establish your trust relationship purely on an owner/maintainer's name at the risk of knowing that if that entity were compromised, your supply chain would be at a larger risk in terms of packages rather than a handful of namespaces. This is somewhat related to the trusted-signers command to add an [author's signature of a package to be trusted](https://docs.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-trusted-signers#options-for-add-based-on-a-package), but without the certificate fingerprint.

If people are interested in being able to trust owners/maintainers in addition to the namespaces they possess, please feel free to upvote & drop some comments below as well. We'll look into it! It may be out of scope initially, but shouldn't stop the conversation.

**Reference:**
https://docs.microsoft.com/en-us/nuget/nuget-org/id-prefix-reservation#id-prefix-reservation-on-nugetorg